Attribute cleanups

nnethercote opened the pull request:
Best reviewed one commit at a time.

r? @petrochenkov

nnethercote committed:
Avoid unnecessary `MetaItem`/`Attribute` conversions.

`check_builtin_attribute` calls `parse_meta` to convert an `Attribute`
to a `MetaItem`, which it then checks. However, many callers of
`check_builtin_attribute` start with a `MetaItem`, and then convert it
to an `Attribute` by calling `cx.attribute(meta_item)`. This `MetaItem`
to `Attribute` to `MetaItem` conversion is silly.

This commit adds a new function `check_builtin_meta_item`, which can be
called instead from these call sites. `check_builtin_attribute` also now
calls it. The commit also renames `check_meta` as `check_attr` to better
match its arguments.

nnethercote committed:
Improve comments about attributes and meta items.

I have found the distinction confusing.

nnethercote committed:
Reorder some types.

So that `Attribute` and `MetaItem` are listed first, and then the
component types are below them in a logical order.

petrochenkov commented on compiler/rustc_builtin_macros/src/derive.rs:
```diff
@@ -33,14 +33,15 @@ impl MultiItemModifier for Expander {
                     template,
                 );
+                let attr =
+                    attr::mk_attr_outer(&sess.parse_sess.attr_id_generator, meta_item.clone());
```
This `attr` is only used for calling `.meta_item_list()` below.
The existing `meta_item` should be enough to do the same thing without creating `attr`.

nnethercote replied:
I tried that, and it turned out to be painful enough that I gave up, I think because it required working at the token level. I could try again, I guess.

nnethercote replied:
You were right, this works, though there was a quirk with an error case in a `ui` test. Details in the commit message in the new commit.

nnethercote committed:
Avoid more unnecessary `MetaItem`/`Attribute` conversions.

In `Expander::expand` the code currently uses `mk_attr_outer` to convert
a `MetaItem` to an `Attribute`, and then follows that with
`meta_item_list` which converts back. This commit avoids the unnecessary
conversions.

There was one wrinkle: the existing conversions caused the bogus `<>` on
`Default<>` to be removed. With the conversion gone, we get a second
error message about the `<>`. This is a rare case, so I think it
probably doesn't matter much.

bors commented:
:umbrella: The latest upstream changes (presumably #105017) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
I rebased.

nnethercote committed:
Remove an out-of-date comment.

This comment dates back to at least 2013, and is no longer relevant.
(There used to be an `allow` attribute, but that's no longer present.)

nnethercote committed:
Inline and remove `MetaItemLit::from_lit_kind`.

It has a single call site.

nnethercote committed:
Avoid more `MetaItem`-to-`Attribute` conversions.

There is code for converting `Attribute` (syntactic) to `MetaItem`
(semantic). There is also code for the reverse direction. The reverse
direction isn't really necessary; it's currently only used when
generating attributes, e.g. in `derive` code.

This commit adds some new functions for creating `Attributes`s directly,
without involving `MetaItem`s: `mk_attr_word`, `mk_attr_name_value_str`,
`mk_attr_nested_word`, and
`ExtCtxt::attr_{word,name_value_str,nested_word}`.

These new methods replace the old functions for creating `Attribute`s:
`mk_attr_inner`, `mk_attr_outer`, and `ExtCtxt::attribute`. Those
functions took `MetaItem`s as input, and relied on many other functions
that created `MetaItems`, which are also removed: `mk_name_value_item`,
`mk_list_item`, `mk_word_item`, `mk_nested_word_item`,
`{MetaItem,MetaItemKind,NestedMetaItem}::token_trees`,
`MetaItemKind::attr_args`, `MetaItemLit::{from_lit_kind,to_token}`,
`ExtCtxt::meta_word`.

Overall this cuts more than 100 lines of code and makes thing simpler.

nnethercote commented:
And now I added three new commits.

petrochenkov commented:
The last commit is something I also wanted to do, there's no need to go through meta-items when generating AST output from built-in macros.
@bors r+

bors commented:
:pushpin: Commit ba1751a201190930a16ef20e6f7021c785d3891d has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov ba1751a201190930a16ef20e6f7021c785d3891d -->
<!-- homu: {"type":"Approved","sha":"ba1751a201190930a16ef20e6f7021c785d3891d","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

nnethercote commented:
I don't think this will affect perf, but the queue is long so there is time to check.

@bors r-

nnethercote commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit ba1751a201190930a16ef20e6f7021c785d3891d with merge a42da13bfac2206c0dd77e74950e6fb81f473449...
<!-- homu: {"type":"TryBuildStarted","head_sha":"ba1751a201190930a16ef20e6f7021c785d3891d","merge_sha":"a42da13bfac2206c0dd77e74950e6fb81f473449"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3578435723/jobs/6020196129)
Build commit: a42da13bfac2206c0dd77e74950e6fb81f473449 (`a42da13bfac2206c0dd77e74950e6fb81f473449`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3578435723/jobs/6020196129"},"merge_sha":"a42da13bfac2206c0dd77e74950e6fb81f473449"} -->

rust-timer commented:
Queued a42da13bfac2206c0dd77e74950e6fb81f473449 with parent e0098a5cc3a87d857e597af824d0ce1ed1ad85e0, future [comparison URL](https://perf.rust-lang.org/compare.html?start=e0098a5cc3a87d857e597af824d0ce1ed1ad85e0&end=a42da13bfac2206c0dd77e74950e6fb81f473449).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (a42da13bfac2206c0dd77e74950e6fb81f473449): [comparison URL](https://perf.rust-lang.org/compare.html?start=e0098a5cc3a87d857e597af824d0ce1ed1ad85e0&end=a42da13bfac2206c0dd77e74950e6fb81f473449&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=e0098a5cc3a87d857e597af824d0ce1ed1ad85e0&end=a42da13bfac2206c0dd77e74950e6fb81f473449&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.1%, -0.1%] | 2     |
| Improvements ✅ <br /> (secondary) | -0.7% | [-0.9%, -0.5%] | 7     |
| All ❌✅ (primary)                 | -0.1% | [-0.1%, -0.1%] | 2     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=e0098a5cc3a87d857e597af824d0ce1ed1ad85e0&end=a42da13bfac2206c0dd77e74950e6fb81f473449&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.0% | [2.0%, 2.0%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=e0098a5cc3a87d857e597af824d0ce1ed1ad85e0&end=a42da13bfac2206c0dd77e74950e6fb81f473449&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.4% | [2.1%, 2.7%] | 4     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

petrochenkov commented:
@bors r+ rollup=maybe

bors commented:
:pushpin: Commit ba1751a201190930a16ef20e6f7021c785d3891d has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov ba1751a201190930a16ef20e6f7021c785d3891d -->
<!-- homu: {"type":"Approved","sha":"ba1751a201190930a16ef20e6f7021c785d3891d","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit ba1751a201190930a16ef20e6f7021c785d3891d with merge d6c4de0fb22fe8f3cc5a27e94b6b6d88fb504a91...
<!-- homu: {"type":"BuildStarted","head_sha":"ba1751a201190930a16ef20e6f7021c785d3891d","merge_sha":"d6c4de0fb22fe8f3cc5a27e94b6b6d88fb504a91"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3590355033/jobs/6046103635)
Approved by: petrochenkov
Pushing d6c4de0fb22fe8f3cc5a27e94b6b6d88fb504a91 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3590355033/jobs/6046103635"},"merge_sha":"d6c4de0fb22fe8f3cc5a27e94b6b6d88fb504a91"} -->

rust-timer commented:
Finished benchmarking commit (d6c4de0fb22fe8f3cc5a27e94b6b6d88fb504a91): [comparison URL](https://perf.rust-lang.org/compare.html?start=744a97b98c3f2722e1a43486c05203b04e499e21&end=d6c4de0fb22fe8f3cc5a27e94b6b6d88fb504a91&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=744a97b98c3f2722e1a43486c05203b04e499e21&end=d6c4de0fb22fe8f3cc5a27e94b6b6d88fb504a91&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.2% | [-0.2%, -0.1%] | 3     |
| Improvements ✅ <br /> (secondary) | -0.7% | [-0.9%, -0.5%] | 6     |
| All ❌✅ (primary)                 | -0.2% | [-0.2%, -0.1%] | 3     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=744a97b98c3f2722e1a43486c05203b04e499e21&end=d6c4de0fb22fe8f3cc5a27e94b6b6d88fb504a91&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.6% | [-0.6%, -0.6%] | 1     |
| Improvements ✅ <br /> (secondary) | -1.0% | [-1.0%, -1.0%] | 1     |
| All ❌✅ (primary)                 | -0.6% | [-0.6%, -0.6%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=744a97b98c3f2722e1a43486c05203b04e499e21&end=d6c4de0fb22fe8f3cc5a27e94b6b6d88fb504a91&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.3%  | [2.3%, 2.3%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.5% | [-1.5%, -1.5%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>
